### PR TITLE
feat (parser.json_v2): Support defining field/tag tables within an object table

### PR DESCRIPTION
**Pull request by sspaink**
resolves: #9381

To allow users to gather specific elements from an array, this pull request allows users to define field/tag tables within an object table. This will allow a user to gather a field/tag while still maintaining the original relationship within the JSON. 

Examples of this new config feature have been added under the testdata folder in the json_v2 directory.

Here is one of the examples to give an idea of the new added feature:

```
[[inputs.file]]
    files = ["./testdata/subfieldtag_in_object/input.json"]
    data_format = "json_v2"
    [[inputs.file.json_v2]]
        [[inputs.file.json_v2.object]]
            path = "root.station"
            [[inputs.file.json_v2.object.field]]
                path = "#.etd.0.estimate.0.minutes"
                type = "int"
            [[inputs.file.json_v2.object.tag]]
                path = "#.abbr"
                rename = "from_station"
            [[inputs.file.json_v2.object.tag]]
                path = "#.etd.0.abbreviation"
                rename = "to_station"
            [[inputs.file.json_v2.object.tag]]
                path = "#.etd.0.estimate.0.direction"
```

### CODE REVIEW

File: plugins/parsers/json_v2/parser_test.go [hunk 1]
```diff
@@ -29,6 +29,18 @@ func TestData(t *testing.T) {
 			name: "Test having an array of objects",
 			test: "array_of_objects",
 		},
+		{
+			name: "test",
```
Comment (sspaink) [inline plugins/parsers/json_v2/parser_test.go:33]: Give this a better name

**Comment by sjwang90** [comment]
Looks good! Ran this config. 

```toml
[[inputs.http]]
    urls = ["http://api.bart.gov/api/etd.aspx?cmd=etd&orig=POWL&key=MW9S-E7SL-26DU-VV8V&json=y"
            ]
    data_format = "json_v2"
    tagexclude = ["url", "host"]
    [[inputs.http.json_v2]]
        [[inputs.http.json_v2.object]]
            path = "root.station"
            [[inputs.http.json_v2.object.tag]]
                path = "#.abbr"
                rename = "from_station"
            # destination = ANTC
            [[inputs.http.json_v2.object.tag]]
                path = "#.etd.0.abbreviation"
                rename = "to_station"
            [[inputs.http.json_v2.object.tag]]
                path = "#.etd.0.estimate.0.direction"
            [[inputs.http.json_v2.object.field]]
                path = "#.etd.0.estimate.0.minutes"
                type = "int"
            # destination = BERY
            [[inputs.http.json_v2.object.tag]]
                path = "#.etd.1.abbreviation"
                rename = "to_station"
            [[inputs.http.json_v2.object.tag]]
                path = "#.etd.1.estimate.0.direction"
            [[inputs.http.json_v2.object.field]]
                path = "#.etd.1.estimate.0.minutes"
                type = "int"
```

with this output
```
> http,etd_estimate_direction=North,from_station=POWL,to_station=ANTC etd_estimate_minutes=9i 1631214122000000000
> http,etd_estimate_direction=North,from_station=POWL,to_station=BERY etd_estimate_minutes=6i 1631214122000000000
```

**Review by sjwang90 — APPROVED** [review]

File: plugins/parsers/json_v2/parser.go [hunk 2]
```diff
@@ -53,21 +62,26 @@ type JSONObject struct {
...
-	OutputName  string
-	SetName     string
-	Tag         bool
-	DesiredType string // Can be "int", "uint", "float", "bool", "string"
+	ParentIndex       int
+	OutputName        string
+	SetName           string
+	Tag               bool
+	DesiredType       string      // Can be "int", "uint", "float", "bool", "string"
+	IncludeCollection *PathResult // If set to true, it should be auto included
```
Comment (helenosheaa) [inline plugins/parsers/json_v2/parser.go:75]: The comment "If set to true" is referring to the bool PathResult.tag or just if IncludeCollection is not nil?

File: plugins/parsers/json_v2/parser.go [hunk 3]
```diff
@@ -314,17 +334,43 @@ func (p *Parser) expandArray(result MetricNode) ([]telegraf.Metric, error) {
...
+					} else {
+						pathResult = p.existsInpathResults(result.ParentIndex, result.Raw)
+					}
+					if pathResult == nil {
+						return results, nil
+					}
+					if pathResult.tag {
+						result.Tag = true
+					}
+
```
Comment (helenosheaa) [inline plugins/parsers/json_v2/parser.go:353]: nit: remove extra line

File: plugins/parsers/json_v2/parser.go [hunk 4]
```diff
@@ -314,17 +334,43 @@ func (p *Parser) expandArray(result MetricNode) ([]telegraf.Metric, error) {
...
+					if !pathResult.tag {
+						desiredType = pathResult.Type
+					}
+
```
Comment (helenosheaa) [inline plugins/parsers/json_v2/parser.go:357]: nit: remove extra line

File: plugins/parsers/json_v2/parser.go [hunk 5]
```diff
@@ -335,22 +381,56 @@ func (p *Parser) expandArray(result MetricNode) ([]telegraf.Metric, error) {
...
 	for _, c := range objects {
 		p.currentSettings = c
+
 		if c.Path == "" {
 			return nil, fmt.Errorf("GJSON path is required")
 		}
-		result := gjson.GetBytes(input, c.Path)
+		result := gjson.GetBytes(p.InputJSON, c.Path)
+
+		// hastag doesn't return index! idea: replace all hastags with index, and find lenght of array
```
Comment (helenosheaa) [inline plugins/parsers/json_v2/parser.go:411]:
```suggestion
		// hashtag doesn't return index! idea: replace all hashtags with index, and find length of array
```

I'm not sure if this comment is supposed to say hashtag or if it's still needed

**Review by helenosheaa — COMMENTED** [review]
Looks good! It's the first time I've properly looked at the new parser, the documentation is awesome. 

Few of nits / questions around certain comments. I'll test out the build as well and leave any feedback if I notice anything, but good job.

File: plugins/parsers/json_v2/parser.go (hunk 5 above)
Comment (sspaink) [reply plugins/parsers/json_v2/parser.go:411]: Your right this shouldn't be here. Removed.

File: plugins/parsers/json_v2/parser.go (hunk 2 above)
Comment (sspaink) [reply plugins/parsers/json_v2/parser.go]: Good catch, I am not sure what I meant by this. It must have been a boolean originally, I've removed it and added a more descriptive comment that hopefully is more clear.

**Comment by sspaink** [comment]
Thanks @helenosheaa for the review, I've addressed the changes and if you could try out the build artifacts that would be great.

**Review by helenosheaa — APPROVED** [review]
tested, looks good!

**Pull request merged**